### ISSUE TITLE
We should avoid to search again from beggining when changing search page.

### ISSUE THREAD

**Issue by mgautierfr** [kiwix/kiwix-tools#97]
From #92 

>Context: every time you click for 25 more search results, it costs you 30-60 seconds or sometimes much more. So serious teachers and serious students would like to see the first 100 results all on one page (or possibly even a much larger number) to avoid bureaucracy, if at all possible!

Every search request in `kiwix-serve` leads to a totally new search.
However, when a user is searching a content and looping through search pages, the search is always the same, only the results range is changing.
We should store the search results and only get new results from the same search instead of creating a new search each time.

It is not a simply thing to do because we need, at least, to store some kind of user session, to associated a search object to a series of search request.

**Comment by MananJethwani** [comment]
@mgautierfr would Redis be helpful in such scenarios?..... It could be used to cache the result.

**Comment by kelson42** [comment]
@MananJethwani you seriously want to make redis a dependence of the libzim or libkiwix?!

**Comment by MananJethwani** [comment]
@kelson42 It's just a suggestion I don't know the whole scope of the project would that be problematic somehow?

**Comment by kelson42** [comment]
@MananJethwani yes, for portability reasons... Ask yourself how to let run redis everywhere the libzim/libkiwix is installed.

**Comment by MananJethwani** [comment]
@kelson42 right sorry as I mentioned I didn't know as well as thought of the whole scope of the project now I think of it you are right we have places where people try to run it on various machines and for various purposes many times it won't be necessary to be kept running.

**Comment by kelson42** [comment]
@mgautierfr A small architectural document about how the new search session/cache in the libzim should work would be helpful at this stage IMO for the GSoC student.

**Comment by mgautierfr** [comment]
It is a multi-project problem.

For now :
- libzim provides only a `Search` class which encapsulate "everything" (A xapian database, a particular query (`Enquire`) and a set of results `MSet`).
- kiwix-lib "simply" forward the argument of the http request to libzim and returns the results. (Differently, but following the same principle for both fulltext search and suggestion)

This means that we somehow redo the same work a lot of time :
- If a user do several search on a zim file (and we do when we use suggestion), we need to recreate a xapian database each time (locate it in the zim file, create a new instance, ...)
- If a user do a fulltext search and browse the different pages, we need to recreate the xapian database and a new (identical) query, just to get a different results set.

To avoid that we need to :
- Change the libzim's search api. I'm currently working on this (PR coming soon). The new api will solve this (among other flaws). The new api will split the `Search` in different class to allow reusing the database (`libzim::Searcher`) and the query itself (`libzim::Search`) allowing to get a set of result `libzim::SearchResult` (the `SearchResult` allowing to iterate over it using iterator).
- kiwix-lib should reuse when possible the search instances (`Searcher`, `Search`) to avoid duplicated work.

This means that kiwix-lib has to keep some instance alive to potentially reuse it later.
This is already the case for the `kiwix::Reader`, we may associate to it a `libzim::Searcher` (at keep both at the same time with same live time).

For suggestion, we probably don't need to keep a `libzim::Search` (next suggestion request will probably be made with different query).
For fulltext search, we need to keep a `libzim::Search` alive to reuse it to get a new `libzim::SearchResult`.
And then, comes the question : "For how long ?"

I have no already made answer for that. But we can think of:
- keep N `Search` alive in a cache (all zim archives and users combined). If there is no room left for a new search in the cache, simply drop the older one.
- keep N `Search` per zim archive.
- keep 1 `Search` per user. Here, we need to identify a user. The simpler solution is to use a cookie (but we need to implement a cookie/session management in kiwix-lib). Then a user may have a `Search` associated with him in the server. If he do a query with the same parameters but the range (change page) we reuse the Search. Else we create a new one and drop the old one. And we have to think to drop `Search` when user leave the site (after a certain time ?)
- anything I haven't think about...

For now, I think the two first solutions are easier to implement and we should start with them. The third one is more complex and not all questions are answered.
But if you have better ideas, I'm open to them.

**Comment by kelson42** [comment]
@mgautierfr And you confirm that all of this should work the same way for both ft and title searches?

**Comment by mgautierfr** [comment]
Almost.

For the suggestion (title search), user will do different search (always with the implicit range 0-10). So we don't need to store the `Search` for later use.
For fulltext search, the pagination system will make the user do the "same" search (using different range). So we need to store the `Search`.

**Comment by maneeshpm** [comment]
@mgautierfr This is a super significant enhancement! This will make search much more manageable. I have an implementation query in the case of multizim. In this case, we can have an overall `Searcher` that contains databases from all the zims and a number of `Searcher`s for individual zims. We might want to keep the overall `Searcher` alive throughout, but what will be the scope of the individual instances, when are they created and destroyed?

**Comment by kelson42** [comment]
> For the suggestion (title search), user will do different search (always with the implicit range 0-10).

I'm not happy with the way it work. I would like to be able to have an auto expanding list of suggestions for example (scrolling down would display more suggestions). There I believe it would be good there too, to do things like for the ft.

**Comment by mgautierfr** [comment]
Yes. Done :)
